generic/MultiMarginCriterion.c

```diff
@@ -127,8 +129,9 @@ static int nn_(MultiMarginCriterion_updateGradInput)(lua_State *L)
     
       if(z > 0)
       {
-        gradInput_target -= g;
-        gradInput_data[d] = g;
+        real h = (p == 1) ? g : 2*g*z;
```

Comment:
same, only works for p=1 or p=2. if this is by design, assert in the constructor. if not, fix it appropriately.